File: 143. Reorder List-1.cpp
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  void reorderList(ListNode *head) {
    if (!head || !head->next)
      return;

    ListNode dummy;
    ListNode *sslow = &dummy;
    sslow->next = head;
    ListNode *fast = head, *slow = head;
    while (fast && fast->next) {
      sslow = sslow->next;
      slow = slow->next;
      fast = fast->next->next;
    }

    sslow->next = nullptr;

    ListNode *prev = nullptr, *cur = slow;
    while (cur) {
      ListNode *nxt = cur->next;
      cur->next = prev;
      prev = cur;
      cur = nxt;
    }

    ListNode *list1 = head;
    ListNode *list2 = prev;
    cur = &dummy;
    while (list1 && list2) {
      cur->next = list1;
      list1 = list1->next;
      cur = cur->next;
      cur->next = list2;
      list2 = list2->next;
      cur = cur->next;
    }
  }
};

File: 143. Reorder List-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        """
        Do not return anything, modify head in-place instead.
        """
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev = None
        cur = slow
        while cur is not None:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt

        head1 = head
        head2 = prev
        while head2.next is not None:
            nxt1 = head1.next
            nxt2 = head2.next
            head1.next = head2
            head2.next = nxt1
            head1 = nxt1
            head2 = nxt2
